tensorflow/core/common_runtime/next_pluggable_device/next_pluggable_device_factory.h

```diff
@@ -41,6 +41,10 @@ class NextPluggableDeviceFactory : public DeviceFactory {
                        const std::string& name_prefix,
                        std::vector<std::unique_ptr<Device>>* devices) override;
 
+  const std::string& compilation_device_name() const {
+    return compilation_device_name_;
```

Comment:
there is a linker error `ld.lld: error: undefined reference due to --no-allow-shlib-undefined: typeinfo for tensorflow::NextPluggableDeviceFactory`. Do you know the cause? Will moving the implementation to next_pluggable_device_factory.cc fix it?

Reply:
I think it may caused by `dynamic_cast`, dynamic_cast needs rtti link option,  I didn't reproduce it yet, does it fail when building android/tf-lite? I saw they will disable rtti in tf-opt
https://stackoverflow.com/questions/307352/g-undefined-reference-to-typeinfo
```
This can also happen when you mix -fno-rtti and -frtti code. Then you need to ensure that any class, which type_info is accessed in the -frtti code, have their key method compiled with -frtti. Such access can happen when you create an object of the class, use dynamic_cast etc.
```
```
182     NextPluggableDeviceFactory* device_factory =
183         dynamic_cast<NextPluggableDeviceFactory*>(
184             DeviceFactory::GetFactory(device_name));
```

Reply:
Got it. Linking some CPU tests (e.g. tensorflow/compiler/jit/xla_platform_info_test_cpu) and GPU tests (e.g. tensorflow/c/env_test_gpu) are failing.

Reply:
@jyingl3  seems the test failure is caused by dynamic_cast<NextPluggableDeviceFactory>, even I removed compilation_device_name.  I modify it as IsPluggableDevice and can create PjRtClient to idendify whether it is a nextpluggabledevice.  seems test passed now. can you help to have a look? thanks
```
  if (DeviceFactory::IsPluggableDevice(device_name) &&
      GetOrCreatePjRtClient(DeviceType(device_name)).ok()) {

```

Reply:
Got it. Can we use GetPjRtClient from tensorflow/core/tfrt/common/pjrt_util.cc instead of GetOrCreatePjRtClient? Just do not want to create a PJRT client unintendedly

Reply:
done